rfcs/react-components/convergence/event-handlers-event-type.md

```diff
@@ -0,0 +1,125 @@
+# RFC: Extending Event Handlers Event Typing
+
+This RFC proposes solutions to enhance event typing within callbacks.
+
+---
+
+@yuanboxue-amber
+
+## Summary
+
+We encounter challenges when attempting to extend event types within callbacks without introducing breaking changes. This RFC offers potential solutions to manage such event type changes for both current stable components and future components.
+
+## Problem statement
+
+Our current design for event callbacks in v9 has the type signature `callback: (e: SomeEvents, data: SomeData) => void;`, where `SomeEvents` represents specific event types like MouseEvent or KeyboardEvent, or a union of specific event types. While our intention is to maintain strong type checking, this design complicates the process of expanding or altering event types without causing breaks.
+
+We experienced this challenge when when adding a scroll event listener to the Popover `onOpenChange` callback. The callback has type `onOpenChange?: (e: OpenPopoverEvents, data: OnOpenChangeData) => void;`, where `OpenPopoverEvents` is a union of specific events. But the scroll event has just type `Event`. And we are facing issues when updating the type:
+
+```ts
+export type OpenPopoverEvents = KeyboardEvent | MouseEvent;
+type Props = {
+  onOpenChange: (e: OpenPopoverEvents) => void;
+};
+const onOpenChange = (e: MouseEvent | KeyboardEvent) => {};
+const props: Props = { onOpenChange }; // 💣 This breaks when we add Event to OpenPopoverEvents. The error: '(e: MouseEvent | KeyboardEvent) => void' is not assignable to type '(e: OpenPopoverEvents) => void'.
+```
+
+This is a challenge especially when unforeseen events need to be added in the future.
+
+## Detailed Design or Proposal
+
+### 1. Solution for existing components in v9:
+
+We cannot introduce generic Event type to the existing callbacks as this is a breaking change for any code that explicitly types its event argument (as in the example in "Problem statement").
+
+#### Solution 1 - cast the new event to the existing type
+
+```ts
+// Within our codebase:
+export type OpenPopoverEvents = KeyboardEvent | MouseEvent;
+
+const usePopover = () => {
+  const handleScroll = (e: Event) => {
+    onOpenChange(e as OpenPopoverEvents, data);
```

Comment:
this is wrong and we should never allow things like this in low level implementation (our code) - we need to guarantee both runtime and ideally type level correctness = embracing defensive programming and adding type guards